### PR TITLE
Fix transfer learning w/ pre-trained LM + finetuning tutorial

### LINKED ISSUE #2964: Initializaing Language model for RNN-Transducer
I am trying to initialise the  RNN-decoder using pre-trained language model with a similar architecture (dlayers, dunits) . 
I get following warning messages : 
```
2021-02-12 13:34:53,311 (asr_init:259) WARNING: model(s) found for pre-initialization
2021-02-12 13:34:53,315 (asr_init:203) WARNING: reading model parameters from /exp/train_rnnlm_pytorch_rnnt4_rnnlm_char/rnnlm.model.20
2021-02-12 13:34:58,336 (asr_init:150) WARNING: module(s) ['att.', ' dec.'] don't match or (partially match) available modules in model.
2021-02-12 13:34:58,337 (asr_init:152) WARNING: for information, the existing modules in model are:
2021-02-12 13:34:58,337 (asr_init:153) WARNING: ['predictor.embed.weight', 'predictor.rnn.0.weight_ih', 'predictor.rnn.0.weight_hh', 'predictor.rnn.0.bias_ih', 'predictor.rnn.0.bias_hh', 'predictor.lo.weight', 'predictor.lo.bias']
2021-02-12 13:34:58,337 (asr_init:243) WARNING: loading [] from model: exp/train_rnnlm_pytorch_rnnt4_rnnlm_char/rnnlm.model.20
e
```
It seems the initalization is not correct, It will be great if somebody can point to sample configuration to do the same.  
Thanks

### PR DESCRIPTION
**Pull request by b-flo**
Closes #2964 + minor refactoring.

Also, It seems some users were confused by the finetuning tutorial so I rewrote the documentation. Please tell me if it's clear now.

### CODE REVIEW

**Comment by sknadig** [comment]
I think the issue mentioned in #2964 is also related to #1757.
This is due to an unnecessary whitespace in [asr_train.py#L383](https://github.com/espnet/espnet/blob/b1753d8397546c0684556504aab75efec9aacb22/espnet/bin/asr_train.py#L383)

```python
In [1]: s = "att., dec."
In [2]: [str(mod) for mod in s.split(",") if s != ""]
Out[2]: ['att.', ' dec.']
```

Fix: Just remove the whitespace between `'att.'` and `'dec.'`
```python
In [3]: s = "att.,dec."
In [4]: [str(mod) for mod in s.split(",") if s != ""]
Out[4]: ['att.', 'dec.']
```

It'd be good if you can add this fix as well. Or I can send a separate PR just for that.

**Comment by b-flo** [comment]
Oh my, thanks! I though I fixed it in a previous PR. It's done now.

Btw, the issue #2964 is in fact due to `filter_module(...)` which was applied before LM key were renamed for ASR task, resulting in empty modules list.

**Comment by sw005320** [comment]
Thanks, @b-flo.
I did not fully check this behavior.
Is there some test?

**Comment by sw005320** [comment]
The documentation looks very good!

**Comment by b-flo** [comment]
> Thanks, @b-flo.
> I did not fully check this behavior.
> Is there some test?

Yes, there are test for each "kind" of transfer learning however I should maybe:
1) compare weights after transfer learning
2) train again after freezing and check weights are not modified

Currently if the user specifies wrong parameters, model path or architecture, etc, we do not stop training. We notify user in training log with the issue, available parameters in target model, etc, and start training with transfer learning disabled.

**Comment by sw005320** [comment]
Thanks!
How about making a mini_an4 integration test with one or two use cases?
This would also give the user a good example.
(sorry to ask many things, but this is one of the most frequent questions, and this will reduce our effort in the future).

We can do it in the later PR if we want to avoid a large PR.

**Comment by b-flo** [comment]
Sure! For now I added integration test for RNN-transducer w/ encoder pre-init (CTC model) and RNN-transducer w/ decoder pre-init (LM). Let me review all cases and tasks using finetuning and I'll add them in another PR if it's okay with you! For freezing option I think it's better to add an unit test to ensure weights are not modified.

Btw, I've just realized that transfer learning for Custom Transducer will not work anymore if other kind of models are used because I'm using different key names. I'll need to add new methods to rename key before transfer.

**Comment by sw005320** [comment]
OK!
If you have some follow-up PRs, you can merge this one by yourself.
My only concern is the test, which can be done later.

**Comment by b-flo** [comment]
> If you have some follow-up PRs, you can merge this one by yourself.
> My only concern is the test, which can be done later.

To be honest, I didn't plan for follow-up PRs yet, this PR was just to fix reported bug and refactor the documentation.
I can put it in stand-by until I add further tests if you want though.

(Btw, I have no idea what's happening with codecov?)

**Comment by sw005320** [comment]
Sorry, I meant to say that “the test code” is the follow up PR.
Some of your PRs tends to be very large and have some difficulties on merging it.
So, I recommend you to merge it frequently...

Also, please do not care so much about the small difference in  the codecov behavior.

**Comment by b-flo** [comment]
> Sorry, I meant to say that “the test code” is the follow up PR.
> Some of your PRs tends to be very large and have some difficulties on merging it.
> So, I recommend you to merge it frequently...

My apologies... I'll split new features/fix containing lot of changes into smaller PRs in the future (after # 2951... :sweat:).

**Comment by sw005320** [comment]
no problem at all. let's focus on 2951 then!

**Comment by lnalborczyk** [comment]
Sorry to jump in, but is there an explicit message in the `train.log` when transfer learning fails and when it is disabled? For instance, in the .log below I receive a few warnings but it is unclear (to me) whether transfer learning is still applied or not?

```
# Started at Wed Feb 24 21:48:47 CET 2021
#
2021-02-24 21:48:52,392 (asr_train:565) WARNING: Skip DEBUG/INFO messages
2021-02-24 21:48:53,387 (asr_init:259) WARNING: model(s) found for pre-initialization
2021-02-24 21:48:53,390 (asr_init:209) WARNING: reading model parameters from /research/crissp/espnet/egs/brainspeak/transfer_mls/exp_large/char_train_no_dev_pytorch_train_mtlalpha0.5/results/model.loss.best
2021-02-24 21:48:53,510 (asr_init:150) WARNING: module(s) [' dec.'] don't match or (partially match) available modules in model.
2021-02-24 21:48:53,511 (asr_init:152) WARNING: for information, the existing modules in model are:
2021-02-24 21:48:53,511 (asr_init:153) WARNING: ['enc.enc.0.birnn0.weight_ih_l0', 'enc.enc.0.birnn0.weight_hh_l0', 'enc.enc.0.birnn0.bias_ih_l0', 'enc.enc.0.birnn0.bias_hh_l0', 'enc.enc.0.birnn0.weight_ih_l0_reverse', 'enc.enc.0.birnn0.weight_hh_l0_reverse', 'enc.enc.0.birnn0.bias_ih_l0_reverse', 'enc.enc.0.birnn0.bias_hh_l0_reverse', 'enc.enc.0.bt0.weight', 'enc.enc.0.bt0.bias', 'enc.enc.0.birnn1.weight_ih_l0', 'enc.enc.0.birnn1.weight_hh_l0', 'enc.enc.0.birnn1.bias_ih_l0', 'enc.enc.0.birnn1.bias_hh_l0', 'enc.enc.0.birnn1.weight_ih_l0_reverse', 'enc.enc.0.birnn1.weight_hh_l0_reverse', 'enc.enc.0.birnn1.bias_ih_l0_reverse', 'enc.enc.0.birnn1.bias_hh_l0_reverse', 'enc.enc.0.bt1.weight', 'enc.enc.0.bt1.bias', 'enc.enc.0.birnn2.weight_ih_l0', 'enc.enc.0.birnn2.weight_hh_l0', 'enc.enc.0.birnn2.bias_ih_l0', 'enc.enc.0.birnn2.bias_hh_l0', 'enc.enc.0.birnn2.weight_ih_l0_reverse', 'enc.enc.0.birnn2.weight_hh_l0_reverse', 'enc.enc.0.birnn2.bias_ih_l0_reverse', 'enc.enc.0.birnn2.bias_hh_l0_reverse', 'enc.enc.0.bt2.weight', 'enc.enc.0.bt2.bias', 'enc.enc.0.birnn3.weight_ih_l0', 'enc.enc.0.birnn3.weight_hh_l0', 'enc.enc.0.birnn3.bias_ih_l0', 'enc.enc.0.birnn3.bias_hh_l0', 'enc.enc.0.birnn3.weight_ih_l0_reverse', 'enc.enc.0.birnn3.weight_hh_l0_reverse', 'enc.enc.0.birnn3.bias_ih_l0_reverse', 'enc.enc.0.birnn3.bias_hh_l0_reverse', 'enc.enc.0.bt3.weight', 'enc.enc.0.bt3.bias', 'ctc.ctc_lo.weight', 'ctc.ctc_lo.bias', 'att.0.mlp_enc.weight', 'att.0.mlp_enc.bias', 'att.0.mlp_dec.weight', 'att.0.mlp_att.weight', 'att.0.loc_conv.weight', 'att.0.gvec.weight', 'att.0.gvec.bias', 'dec.embed.weight', 'dec.decoder.0.weight_ih', 'dec.decoder.0.weight_hh', 'dec.decoder.0.bias_ih', 'dec.decoder.0.bias_hh', 'dec.output.weight', 'dec.output.bias', 'dec.att.0.mlp_enc.weight', 'dec.att.0.mlp_enc.bias', 'dec.att.0.mlp_dec.weight', 'dec.att.0.mlp_att.weight', 'dec.att.0.loc_conv.weight', 'dec.att.0.gvec.weight', 'dec.att.0.gvec.bias']
2021-02-24 21:48:53,511 (asr_init:243) WARNING: loading ['att.'] from model: /research/crissp/espnet/egs/brainspeak/transfer_mls/exp_large/char_train_no_dev_pytorch_train_mtlalpha0.5/results/model.loss.best
2021-02-24 21:48:53,511 (asr_init:246) WARNING: override att.0.mlp_enc.weight
2021-02-24 21:48:53,511 (asr_init:246) WARNING: override att.0.mlp_enc.bias
2021-02-24 21:48:53,511 (asr_init:246) WARNING: override att.0.mlp_dec.weight
2021-02-24 21:48:53,511 (asr_init:246) WARNING: override att.0.mlp_att.weight
2021-02-24 21:48:53,511 (asr_init:246) WARNING: override att.0.loc_conv.weight
2021-02-24 21:48:53,511 (asr_init:246) WARNING: override att.0.gvec.weight
2021-02-24 21:48:53,511 (asr_init:246) WARNING: override att.0.gvec.bias
```

NB: I have specified the decoder initialisation with: dec-init: /research/crissp/espnet/egs/brainspeak/transfer_mls/exp_large/char_train_no_dev_pytorch_train_mtlalpha0.5/results/model.loss.best

**Comment by b-flo** [comment]
> Sorry to jump in, but is there an explicit message in the train.log when transfer learning fails and when it is disabled? For instance, in the .log below I receive a few warnings but it is unclear (to me) whether transfer learning is still applied or not?

If any module weight is transferred, transfer learning won't be disabled even if there are some mismatch for other modules.

> From what I understand of the above log file, only the attention weights are taken from the pretrained model, is that correct?

It's correct!

**Comment by lnalborczyk** [comment]
Thank you for your prompt answer!

Also, I often get this kind of warning:

`2021-02-26 11:55:03,370 (asr_init:287) WARNING: modules ['att.', 'dec.'] in model /research/crissp/espnet/egs/brainspeak/transfer_mls/exp_large/char_train_no_dev_pytorch_train_mtlalpha0.5/results/model.loss.best don't match your training config`

Is there a way of knowing more precisely which config parameters do not match?

**Comment by b-flo** [comment]
From what I remember, the mismatched modules are reported in logs alongside transferred modules and complete list of modules in both models? If not, sorry, I have not followed the changes made in related methods.
Based on your feedback I'll modify that part when I have time.

Btw, you could compare pre-trained `model.json` (or training config) and target `model.json` (or training config) and easily spot difference for the modules parameters you selected during transfer.
